joomla/joomla-cms
# [5.1] Form: Add regex validation rule

### Summary of Changes
This PR adds a regex validation rule for the Form library. The base class for Form validation rules has validation with regular expression implemented, but no way to hand in a regular expression via the form definition. You have to build your own validation rule class, register it in Joomla and can then use it. That seems overly complex for this, so this PR adds a rule to add regular expression validation.

For your field, you add `validate="regex"` to your form definition and the regular expression via `validate_regex=""`. The value of the field is then validated server side against the regex given. You have to pay attention that you might have to escape the regex properly when adding it to the validate_regex attribute. You can also add a modifier at the end.


### Testing Instructions
1. Create a new instance of a core module, for example mod_custom in the backend and go to the "Advanced" tab. Modify the value of one of the entries in the "Layout" select to be a value which does not contain A-Z, a number, - or _. Invalid would for example be `üüü`.
2. When saving, you should get an error, that this is invalid.
3. Edit /modules/mod_custom/mod_custom.xml and replace `validate="moduleLayout"` with 
	```
					validate="regex"
					validate_regex="^([A-Za-z0-9_-]+:)?[A-Za-z0-9-][A-Za-z0-9\.-]*$"
	```
4. Try the same as in 1. and see that it still fails, but that the other values still properly pass.

Or it could be approved by maintainers by codereview.


### Thanks
A big thanks goes to @coolcat-creations ([Website](https://coolcat-creations.com/)) who sponsored this PR.

### Link to documentations
Please select:

- [X] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/228
